File: frontend/src/services/graphBuilder.ts
import type { GraphEdge, GraphEdgeType, GraphNode, GraphResponse } from "../types/graph";
import {
  getBranches,
  getBranchCommits,
  parseGitHubRepoUrl,
  type RepositoryRef,
  type GitHubBranch,
  type GitHubCommit
} from "./github";
import { initialVisibleBranches } from "./branchSelection";

interface RepositoryGraphCache {
  repo: RepositoryRef;
  branches: GitHubBranch[];
  commitsByBranch: Map<string, GitHubCommit[]>;
  maxCommits: number;
}

const graphCaches = new Map<string, RepositoryGraphCache>();

function commitId(branchName: string, sha: string): string {
  return `commit:${branchName}:${sha}`;
}

function commitNode(commit: GitHubCommit, branchName: string): GraphNode {
  return {
    id: commitId(branchName, commit.sha),
    label: commit.sha.slice(0, 4),
    type: "commit",
    branch: branchName,
    sha: commit.sha,
    parentShas: (commit.parents ?? []).flatMap((parent) => parent.sha ? [parent.sha] : []),
    author: commit.commit?.author?.name ?? null,
    date: commit.commit?.author?.date ?? null,
    message: commit.commit?.message?.split("\n")[0] ?? "",
    url: commit.html_url ?? null
  };
}

function branchNode(branchName: string): GraphNode {
  return {
    id: `branch:${branchName}`,
    label: branchName,
    type: "branch",
    branch: branchName
  };
}

function branchPriority(branch: GitHubBranch): [number, string] {
  const name = branch.name.toLowerCase();
  if (["development", "develop", "dev"].includes(name)) {
    return [1, name];
  }
  if (name === "qa") {
    return [2, name];
  }
  if (["main", "master"].includes(name)) {
    return [3, name];
  }
  return [0, name];
}

function compareBranches(left: GitHubBranch, right: GitHubBranch): number {
  const [leftPriority, leftName] = branchPriority(left);
  const [rightPriority, rightName] = branchPriority(right);
  return leftPriority - rightPriority || leftName.localeCompare(rightName);
}

function isPrimaryEnvironmentBranch(branchName: string): boolean {
  return ["development", "develop", "dev", "qa"].includes(branchName.toLowerCase());
}

function branchProject(branchName: string): string {
  const separatorIndex = branchName.indexOf("/");
  return separatorIndex > 0 ? branchName.slice(0, separatorIndex).toLowerCase() : "";
}

function lastBranchSegment(branchName: string): string {
  return branchName.split("/").pop()?.toLowerCase() ?? branchName.toLowerCase();
}

function isBaseBranch(branchName: string): boolean {
  return [
    "main",
    "master",
    "development",
    "dev",
    "qa",
    "staging",
    "prod",
    "production"
  ].includes(lastBranchSegment(branchName));
}

function isChildBranch(branchName: string): boolean {
  return branchName
    .toLowerCase()
    .split("/")
    .some((segment) => ["feature", "bugfix", "hotfix", "release", "task", "fix"].includes(segment));
}

function mergeDestinationRank(branchName: string): number {
  if (isBaseBranch(branchName)) {
    return 0;
  }
  if (isChildBranch(branchName)) {
    return 2;
  }
  return 1;
}

function addEdge(edges: Map<string, GraphEdge>, edge: GraphEdge): void {
  if (!edges.has(edge.id)) {
    edges.set(edge.id, edge);
  }
}

function branchFromNode(node: GraphNode | undefined): string | null {
  return node?.branch ?? null;
}

function hasBranchEvidence(
  edges: Map<string, GraphEdge>,
  nodes: Map<string, GraphNode>,
  leftBranch: string,
  rightBranch: string
): boolean {
  return Array.from(edges.values()).some((edge) => {
    const sourceBranch = branchFromNode(nodes.get(edge.source));
    const targetBranch = branchFromNode(nodes.get(edge.target));
    return (
      (sourceBranch === leftBranch && targetBranch === rightBranch)
      || (sourceBranch === rightBranch && targetBranch === leftBranch)
    );
  });
}

function hasIncomingBranchInference(edges: Map<string, GraphEdge>, branchName: string): boolean {
  return Array.from(edges.values()).some((edge) => (
    edge.target === `branch:${branchName}`
    && (edge.type === "branch_possible" || edge.type === "branch_assumed")
  ));
}

function commitSetKey(commits: GitHubCommit[]): string | null {
  if (commits.length === 0) {
    return null;
  }

  return commits.map((commit) => commit.sha).sort().join("|");
}

function hasSharedHistory(
  parentBranch: string,
  childBranch: string,
  branchCommitShas: Map<string, Set<string>>
): boolean {
  const parentShas = branchCommitShas.get(parentBranch);
  const childShas = branchCommitShas.get(childBranch);
  if (!parentShas || !childShas) {
    return false;
  }

  return Array.from(parentShas).some((sha) => childShas.has(sha));
}

/** Chooses the branch that should own a merge commit when multiple visible branches contain it. */
function preferredMergeDestination(
  mergeSha: string,
  sourceBranch: string,
  branchCommitShas: Map<string, Set<string>>,
  branchCommitIndexes: Map<string, Map<string, number>>,
  branchPositions: Map<string, number>
): string | null {
  const sourceProject = branchProject(sourceBranch);
  const candidates = Array.from(branchCommitShas.entries())
    .filter(([branchName, shas]) => branchName !== sourceBranch && shas.has(mergeSha))
    .map(([branchName]) => branchName);

  return candidates
    .sort((left, right) => {
      const leftSameProject = branchProject(left) === sourceProject ? 0 : 1;
      const rightSameProject = branchProject(right) === sourceProject ? 0 : 1;
      return (
        leftSameProject - rightSameProject
        || mergeDestinationRank(left) - mergeDestinationRank(right)
        || (branchCommitIndexes.get(left)?.get(mergeSha) ?? Number.POSITIVE_INFINITY)
        - (branchCommitIndexes.get(right)?.get(mergeSha) ?? Number.POSITIVE_INFINITY)
        || (branchPositions.get(left) ?? 0) - (branchPositions.get(right) ?? 0)
        || left.localeCompare(right)
      );
    })[0] ?? null;
}

function graphCacheKey(repo: RepositoryRef, maxCommits: number): string {
  return `${repo.fullName.toLowerCase()}:${maxCommits}`;
}

async function loadMissingBranchCommits(
  cache: RepositoryGraphCache,
  branchNames: string[]
): Promise<void> {
  const availableBranches = new Set(cache.branches.map((branch) => branch.name));
  const missingBranches = Array.from(new Set(branchNames))
    .filter((branchName) => availableBranches.has(branchName) && !cache.commitsByBranch.has(branchName));

  await Promise.all(
    missingBranches.map(async (branchName) => {
      const commits = await getBranchCommits(cache.repo, branchName, cache.maxCommits);
      cache.commitsByBranch.set(branchName, commits);
    })
  );
}

/** Builds graph nodes and direct Git/PR edges from the cached GitHub branch data. */
function graphFromCache(cache: RepositoryGraphCache): GraphResponse {
  const { repo, branches, commitsByBranch: cachedCommits } = cache;
  if (branches.length === 0) {
    return { repository: repo.fullName, nodes: [], edges: [] };
  }

  const nodes = new Map<string, GraphNode>();
  const edges = new Map<string, GraphEdge>();
  const branchHeads = new Map<string, string>();
  const branchCommitShas = new Map<string, Set<string>>();
  const branchCommitIndexes = new Map<string, Map<string, number>>();
  const branchCommits = new Map<string, GitHubCommit[]>();
  const branchPositions = new Map<string, number>();

  for (const [branchIndex, branch] of branches.entries()) {
    const branchName = branch.name;
    branchPositions.set(branchName, branchIndex);
    const branchGraphNode = branchNode(branchName);
    nodes.set(branchGraphNode.id, branchGraphNode);

    const commits = cachedCommits.get(branchName) ?? [];
    branchCommits.set(branchName, commits);
    const commitShas = new Set(commits.map((commit) => commit.sha));
    const commitIndexes = new Map(commits.map((commit, index) => [commit.sha, index]));
    branchCommitShas.set(branchName, commitShas);
    branchCommitIndexes.set(branchName, commitIndexes);
    if (commits.length > 0) {
      branchHeads.set(branchName, commits[0].sha);
    }

    commits.forEach((commit, index) => {
      const id = commitId(branchName, commit.sha);
      nodes.set(id, commitNode(commit, branchName));

      if (index === 0) {
        addEdge(edges, {
          id: `branch-head:${branchName}:${commit.sha}`,
          source: `branch:${branchName}`,
          target: id,
          type: "branch_commit",
          branch: branchName
        });
      }

      const directParent = (commit.parents ?? [])
        .filter((parent): parent is { sha: string } => (
          Boolean(parent.sha)
          && commitShas.has(parent.sha as string)
          && (commitIndexes.get(parent.sha as string) ?? -1) > index
        ))
        .sort((left, right) => (
          (commitIndexes.get(left.sha) ?? Number.POSITIVE_INFINITY)
          - (commitIndexes.get(right.sha) ?? Number.POSITIVE_INFINITY)
        ))[0];

      if (directParent) {
        const type: GraphEdgeType = (commit.parents?.length ?? 0) > 1 ? "merge" : "parent";
        addEdge(edges, {
          id: `${type}:${branchName}:${commit.sha}:${directParent.sha}`,
          source: id,
          target: commitId(branchName, directParent.sha),
          type,
          branch: branchName
        });
      }
    });
  }

  const representedMerges = new Set<string>();
  // Merge edges are created only from direct GitHub parent metadata, never from transitive ancestry.
  branchCommits.forEach((commits, destinationBranch) => {
    commits.forEach((mergeCommit) => {
      const parents = mergeCommit.parents ?? [];
      if (parents.length < 2 || representedMerges.has(mergeCommit.sha)) {
        return;
      }

      for (const mergedParent of parents.slice(1)) {
        if (!mergedParent.sha) {
          continue;
        }
        const sourceCandidates = Array.from(branchCommitShas.entries())
          .filter(([branchName, shas]) => (
            branchName !== destinationBranch
            && shas.has(mergedParent.sha as string)
            // A downstream branch may inherit both commits. It is not the direct merge source.
            && !shas.has(mergeCommit.sha)
          ))
          .map(([branchName]) => branchName);
        if (sourceCandidates.length === 0) {
          continue;
        }
        const destinationPosition = branchPositions.get(destinationBranch) ?? 0;
        const earlierCandidates = sourceCandidates.filter((branchName) => (
          (branchPositions.get(branchName) ?? 0) < destinationPosition
        ));
        const sourceBranch = (earlierCandidates.length > 0 ? earlierCandidates : sourceCandidates)
          .sort((left, right) => (
            (branchPositions.get(right) ?? 0) - (branchPositions.get(left) ?? 0)
          ))[0];
        if (preferredMergeDestination(
          mergeCommit.sha,
          sourceBranch,
          branchCommitShas,
          branchCommitIndexes,
          branchPositions
        ) !== destinationBranch) {
          continue;
        }

        addEdge(edges, {
          id: `pull-request-merge:${sourceBranch}:${destinationBranch}:${mergeCommit.sha}`,
          source: commitId(sourceBranch, mergedParent.sha),
          target: commitId(destinationBranch, mergeCommit.sha),
          type: "pull_request_merge",
          branch: sourceBranch
        });
        representedMerges.add(mergeCommit.sha);
        break;
      }
    });
  });

  const branchesByCommitSet = new Map<string, string[]>();
  branchCommits.forEach((commits, branchName) => {
    const key = commitSetKey(commits);
    if (!key) {
      return;
    }

    branchesByCommitSet.set(key, [...(branchesByCommitSet.get(key) ?? []), branchName]);
  });

  // Identical histories need a deterministic branch-level tie-breaker without changing commit edges.
  branchesByCommitSet.forEach((matchingBranches) => {
    if (matchingBranches.length < 2) {
      return;
    }

    const bases = matchingBranches.filter(isBaseBranch);
    if (bases.length === 0) {
      return;
    }

    matchingBranches
      .filter((branchName) => isChildBranch(branchName))
      .forEach((childBranch) => {
        const childProject = branchProject(childBranch);
        const parentBranch = bases.find((baseBranch) => (
          baseBranch !== childBranch
          && branchProject(baseBranch) === childProject
          && !hasBranchEvidence(edges, nodes, baseBranch, childBranch)
        ));

        if (!parentBranch) {
          return;
        }

        addEdge(edges, {
          id: `branch_possible:identical-history:${parentBranch}:${childBranch}`,
          source: `branch:${parentBranch}`,
          target: `branch:${childBranch}`,
          type: "branch_possible",
          branch: childBranch
        });
      });
  });

  const parentBranch = Array.from(branchHeads.keys()).find((branch) => branch.toLowerCase() === "main")
    ?? Array.from(branchHeads.keys()).find((branch) => branch.toLowerCase() === "master");

  if (parentBranch) {
    const parentHead = branchHeads.get(parentBranch) as string;
    branchHeads.forEach((childHead, childBranch) => {
      if (childBranch === parentBranch) {
        return;
      }
      if (hasIncomingBranchInference(edges, childBranch)) {
        return;
      }

      let type: GraphEdgeType = "branch_possible";
      const sharesParentHistory = hasSharedHistory(parentBranch, childBranch, branchCommitShas);
      if (isPrimaryEnvironmentBranch(childBranch)) {
        if (childHead === parentHead) {
          type = "branch_assumed";
        } else if (!sharesParentHistory) {
          return;
        }
      } else if (!sharesParentHistory) {
        return;
      }

      addEdge(edges, {
        id: `${type}:${parentBranch}:${childBranch}`,
        source: `branch:${parentBranch}`,
        target: `branch:${childBranch}`,
        type,
        branch: childBranch
      });
    });
  }

  return {
    repository: repo.fullName,
    nodes: Array.from(nodes.values()),
    edges: Array.from(edges.values())
  };
}

/** Loads the default branch set and returns the initial repository graph. */
export async function buildRepositoryGraph(repoUrl: string, maxCommits: number): Promise<GraphResponse> {
  const repo = parseGitHubRepoUrl(repoUrl);
  const branches = (await getBranches(repo)).sort(compareBranches);
  const key = graphCacheKey(repo, maxCommits);
  const previousCache = graphCaches.get(key);
  const cache: RepositoryGraphCache = {
    repo,
    branches,
    commitsByBranch: previousCache?.commitsByBranch ?? new Map(),
    maxCommits
  };

  graphCaches.set(key, cache);
  const initialBranches = initialVisibleBranches(branches.map((branch) => branch.name));
  initialBranches.forEach((branchName) => cache.commitsByBranch.delete(branchName));
  await loadMissingBranchCommits(cache, initialBranches);
  return graphFromCache(cache);
}

/** Loads additional branch histories into the existing repository graph cache. */
export async function loadRepositoryBranches(
  repoUrl: string,
  maxCommits: number,
  branchNames: string[]
): Promise<GraphResponse> {
  const repo = parseGitHubRepoUrl(repoUrl);
  const cache = graphCaches.get(graphCacheKey(repo, maxCommits));
  if (!cache) {
    return buildRepositoryGraph(repoUrl, maxCommits);
  }

  await loadMissingBranchCommits(cache, branchNames);
  return graphFromCache(cache);
}

/** Refreshes the selected visible branches while preserving the graph cache contract. */
export async function refreshRepositoryBranches(
  repoUrl: string,
  maxCommits: number,
  branchNames: string[]
): Promise<GraphResponse> {
  const repo = parseGitHubRepoUrl(repoUrl);
  const branches = (await getBranches(repo)).sort(compareBranches);
  const key = graphCacheKey(repo, maxCommits);
  const previousCache = graphCaches.get(key);
  const cache: RepositoryGraphCache = {
    repo,
    branches,
    commitsByBranch: previousCache?.commitsByBranch ?? new Map(),
    maxCommits
  };

  graphCaches.set(key, cache);
  branchNames.forEach((branchName) => cache.commitsByBranch.delete(branchName));
  await loadMissingBranchCommits(cache, branchNames);
  return graphFromCache(cache);
}

File: backend/app/services/graph_builder.py
from app.models import GraphEdge, GraphNode, GraphResponse
from app.services.github import GitHubClient, RepositoryRef


def _commit_label(sha: str) -> str:
    return sha[:4]


def _commit_id(branch_name: str, sha: str) -> str:
    return f"commit:{branch_name}:{sha}"


def _commit_node(commit: dict, branch_name: str) -> GraphNode:
    sha = commit["sha"]
    commit_data = commit.get("commit", {})
    author = commit_data.get("author") or {}
    message = (commit_data.get("message") or "").splitlines()[0]

    return GraphNode(
        id=_commit_id(branch_name, sha),
        label=_commit_label(sha),
        type="commit",
        branch=branch_name,
        author=author.get("name"),
        date=author.get("date"),
        message=message,
        url=commit.get("html_url"),
    )


def _branch_node(branch_name: str) -> GraphNode:
    return GraphNode(
        id=f"branch:{branch_name}",
        label=branch_name,
        type="branch",
        branch=branch_name,
    )

# Las ramas base quedan al final para que main sea el origen visual de las derivadas.
def _branch_sort_key(branch: dict) -> tuple[int, str]:
    name = branch["name"].lower()
    if name in {"development", "develop", "dev"}:
        return (1, name)
    if name == "qa":
        return (2, name)
    if name in {"main", "master"}:
        return (3, name)
    return (0, name)


def _is_primary_environment_branch(branch_name: str) -> bool:
    return branch_name.lower() in {"development", "develop", "dev", "qa"}


async def build_repository_graph(
    github: GitHubClient,
    repo: RepositoryRef,
    max_commits: int,
) -> GraphResponse:
    branches = sorted(await github.get_branches(repo), key=_branch_sort_key)
    if not branches:
        return GraphResponse(repository=repo.full_name, nodes=[], edges=[])

    nodes_by_id: dict[str, GraphNode] = {}
    edges_by_id: dict[str, GraphEdge] = {}
    branch_heads: dict[str, str] = {}
    branch_commit_shas: dict[str, set[str]] = {}
    branch_commits: dict[str, list[dict]] = {}
    branch_positions: dict[str, int] = {}

    for branch_index, branch in enumerate(branches):
        branch_name = branch["name"]
        branch_positions[branch_name] = branch_index
        branch_node = _branch_node(branch_name)
        nodes_by_id[branch_node.id] = branch_node

        commits = await github.get_branch_commits(repo, branch_name, max_commits)
        branch_commits[branch_name] = commits
        commit_shas = {commit["sha"] for commit in commits}
        branch_commit_shas[branch_name] = commit_shas
        if commits:
            branch_heads[branch_name] = commits[0]["sha"]

        for index, commit in enumerate(commits):
            sha = commit["sha"]
            commit_id = _commit_id(branch_name, sha)

            nodes_by_id[commit_id] = _commit_node(commit, branch_name)

            if index == 0:
                edge = GraphEdge(
                    id=f"branch-head:{branch_name}:{sha}",
                    source=f"branch:{branch_name}",
                    target=commit_id,
                    type="branch_commit",
                    branch=branch_name,
                )
                edges_by_id[edge.id] = edge

            for parent in commit.get("parents", []):
                parent_sha = parent.get("sha")
                if not parent_sha or parent_sha not in commit_shas:
                    continue
                parent_id = _commit_id(branch_name, parent_sha)

                edge_type = "merge" if len(commit.get("parents", [])) > 1 else "parent"
                edge = GraphEdge(
                    id=f"{edge_type}:{branch_name}:{sha}:{parent_sha}",
                    source=commit_id,
                    target=parent_id,
                    type=edge_type,
                    branch=branch_name,
                )
                edges_by_id.setdefault(edge.id, edge)

    represented_merges: set[str] = set()
    for destination_branch, commits in branch_commits.items():
        for merge_commit in commits:
            merge_sha = merge_commit["sha"]
            parents = merge_commit.get("parents", [])
            if len(parents) < 2 or merge_sha in represented_merges:
                continue

            for merged_parent in parents[1:]:
                merged_sha = merged_parent.get("sha")
                if not merged_sha:
                    continue

                source_candidates = [
                    branch_name
                    for branch_name, shas in branch_commit_shas.items()
                    if branch_name != destination_branch and merged_sha in shas
                ]
                if not source_candidates:
                    continue

                earlier_candidates = [
                    branch_name
                    for branch_name in source_candidates
                    if branch_positions[branch_name] < branch_positions[destination_branch]
                ]
                source_branch = max(
                    earlier_candidates or source_candidates,
                    key=lambda branch_name: branch_positions[branch_name],
                )
                edge = GraphEdge(
                    id=f"pull-request-merge:{source_branch}:{destination_branch}:{merge_sha}",
                    source=_commit_id(source_branch, merged_sha),
                    target=_commit_id(destination_branch, merge_sha),
                    type="pull_request_merge",
                    branch=source_branch,
                )
                edges_by_id[edge.id] = edge
                represented_merges.add(merge_sha)
                break

    parent_branch = next(
        (branch_name for branch_name in branch_heads if branch_name.lower() == "main"),
        next((branch_name for branch_name in branch_heads if branch_name.lower() == "master"), None),
    )

    if parent_branch:
        parent_head = branch_heads[parent_branch]
        for child_branch, child_head in branch_heads.items():
            if child_branch == parent_branch:
                continue

            if _is_primary_environment_branch(child_branch):
                if child_head == parent_head:
                    edge_type = "branch_assumed"
                elif parent_head in branch_commit_shas[child_branch]:
                    edge_type = "branch_possible"
                else:
                    continue
            else:
                edge_type = "branch_possible"

            edge = GraphEdge(
                id=f"{edge_type}:{parent_branch}:{child_branch}",
                source=f"branch:{parent_branch}",
                target=f"branch:{child_branch}",
                type=edge_type,
                branch=child_branch,
            )
            edges_by_id[edge.id] = edge

    return GraphResponse(
        repository=repo.full_name,
        nodes=list(nodes_by_id.values()),
        edges=list(edges_by_id.values()),
    )
